File: scraper/shopclues.py
from bs4 import BeautifulSoup
from lxml import etree
import requests

def process_product(product, priceList):

    try:
        price = product.xpath('.//span[@class="p_price"]')[0].text.replace('₹', '').replace(',', '').strip()
        if price in priceList:
            return None
        priceList.append(price)

        link = "https:" + str(product.xpath('.//a')[0].get('href'))
        name = product.xpath('.//h2')[0].text
        seller = "ShopClues"
        img = str(product.xpath(".//img[1]")[0].get("data-img"))
        freeDelivery = True if product.xpath("//div[@class='shipping']//span")[0].text == "Free Shipping" else False

        return {
            "name": name,
            "price": int(price),
            "rating": 0,
            "seller": seller,
            "img": img,
            "text": '',
            "link": link,
            'freeDelivery': freeDelivery
        }
    except:
        return {
            "name": "",
            "price": 0,
            "rating": 0.0,
            "seller": "",
            "img": "",
            "text": "",
            "link": "",
            'freeDelivery': False,
        }

def search(q):
    url = "https://www.shopclues.com/search?q=" + q
    response = requests.get(url).content
    soup = BeautifulSoup(response, 'lxml')
    dom = etree.HTML(str(soup))
    products = dom.xpath('//*[contains(concat( " ", @class, " " ), concat( " ", "search_blocks", " " ))]')

    priceList = []
    result = []

    for product in products:
        product_data = process_product(product, priceList)
        if product_data is not None:
            result.append(product_data)

    return result

File: scraper/shopclues.js
import {load} from 'cheerio';
import bent from 'bent';

const shopclues = async (q) => {
    const get = bent('https://www.shopclues.com/search?q=', 'GET', 'string', 200);
    const response = await get(q);
    const $ = load(response);
    const data = [];

    $('.column.col3.search_blocks').each((index, element) => {
        const title = $(element).find('h2').text();
        if (!title || title == '') return;
        const link = $(element).find('a').attr('href');
        if (!link || link == '') return;
        const rating = 404;
        const price = parseInt($(element).find('.p_price').text().match(/\d+/g).join(''));
        if (!price  || price == '') return;
        let freeDelivery = false;
        if ($(element).find('.shipping').text().toLowerCase().includes("free")){
            freeDelivery = true;
        }
        const desc = $(element).find('.prd_discount').text();
        const img = $(element).find('img').attr('src');
        const trust = false;
        
        data.push({ 
            "title": title,
            "link": `https:${link}`,
            "rating": rating,
            "price": price,
            "freeDelivery": freeDelivery,
            "desc": desc,
            "img": img,
            "trust": trust
        });
    });
    return data;
}

export default shopclues;